Index: transmission-1.82/third-party/Makefile.am
===================================================================
--- transmission-1.82.orig/third-party/Makefile.am	2010-01-23 16:11:29.000000000 +0100
+++ transmission-1.82/third-party/Makefile.am	2010-01-23 18:17:09.000000000 +0100
@@ -1,4 +1,4 @@
-SUBDIRS = libnatpmp miniupnp libevent dht
+SUBDIRS = libnatpmp miniupnp dht
 
 EXTRA_DIST = \
   macosx-libevent-config.h \
Index: transmission-1.82/configure.ac
===================================================================
--- transmission-1.82.orig/configure.ac	2010-01-23 18:13:20.000000000 +0100
+++ transmission-1.82/configure.ac	2010-01-23 18:19:23.000000000 +0100
@@ -190,7 +190,7 @@ AC_CHECK_LIB([rt],
 dnl build our copy of libevent whether we use it or not,
 dnl because "make dist" needs its Makefiles to exist, and
 dnl AM_CONDITIONAL + AC_CONFIG_SUBDIRS don't seem to play nice
-AC_CONFIG_SUBDIRS([third-party/libevent])
+#AC_CONFIG_SUBDIRS([third-party/libevent])
 libevent_source=bundled
 AC_CHECK_LIB([event],[evutil_vsnprintf],
              [libevent_found=yes],
